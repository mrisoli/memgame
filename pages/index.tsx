import Head from 'next/head'

export const Home = (): JSX.Element => (
  <div className="container">
    <Head>
      <title>Memory game</title>
      <link rel="icon" href="/favicon.ico" />
    </Head>

    <main>
      <div>Board</div>
      <div>Status</div>
      <div>Reset</div>
    </main>

    <style jsx>{`
    `}</style>
  </div>
)

export default Home
