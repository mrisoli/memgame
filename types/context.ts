export enum BoardActionType {
  RESET = 'RESET',
  SELECT_CARD = 'SELECT_CARD',
}

export type BoardAction = {type: BoardActionType.RESET, count?: number}
  | {type: BoardActionType.SELECT_CARD, card: number }
export type BoardDispatch = (action: BoardAction) => void
export type BoardState = {
  active: number[]
  board: number[]
  matched: Set<number>
}
