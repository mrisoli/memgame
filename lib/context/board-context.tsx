import { createContext, FC, useContext, useReducer } from 'react'
import { GameStatus } from '../../types'
import { BoardAction, BoardActionType, BoardDispatch, BoardState } from '../../types/context'
import { checkGame, checkMatch, getInitialBoard } from '../utils'

const BoardStateContext = createContext<BoardState | undefined>(undefined)
const BoardDispatchContext = createContext<BoardDispatch | undefined>(undefined)

const boardReducer = (state: BoardState, action: BoardAction) => {
  switch(action.type) {
    case BoardActionType.CHECK_MATCH:
      return {...state, gameStatus: GameStatus.MATCHING, matched: checkMatch(state)}
    case BoardActionType.FINISH_GAME:
      return {...state, gameStatus: GameStatus.WON }
    case BoardActionType.RESET:
      return getInitialBoard(action.count)
    case BoardActionType.SELECT_CARD:
      return {...state, active: [...state.active, action.key]}
    case BoardActionType.STOP_MATCHING:
      return { ...state, active: [], gameStatus: checkGame(state.gameStatus) }
  }
}

const BoardProvider: FC = ({children}) => {
  const [state, dispatch] = useReducer(boardReducer, getInitialBoard(6))

  return (
    <BoardStateContext.Provider value={state}>
      <BoardDispatchContext.Provider value={dispatch}>
        {children}
      </BoardDispatchContext.Provider>
    </BoardStateContext.Provider>
  )
}

const useBoardState: () => BoardState = () => {
  const context = useContext(BoardStateContext)
  if (context === undefined) {
    throw new Error('useBoardState must be used within a BoardProvider')
  }
  return context
}

const useBoardDispatch: () => BoardDispatch = () => {
  const context = useContext(BoardDispatchContext)
  if (context === undefined) {
    throw new Error('useBoardDispatch must be used within a BoardProvider')
  }
  return context
}

export { BoardProvider, useBoardDispatch, useBoardState }
