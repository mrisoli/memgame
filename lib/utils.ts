import { GameStatus } from "../types"
import { BoardState } from "../types/context"

export const checkGame: (gameStatus: GameStatus) => GameStatus =
  (gameStatus) => gameStatus == GameStatus.WON ? GameStatus.WON : GameStatus.IN_PROGRESS

export const checkMatch: (state: BoardState) => Set<number> =
  ({ active, board, matched }) => {
  const [first, second] = active.map(key => board[key])
  return first == second ? matched.add(first) : matched
}


export const getInitialBoard: (count: number) => BoardState = (count) => ({
  active: [],
  board: resetBoard(count),
  gameStatus: GameStatus.IN_PROGRESS,
  matched: new Set()
})

export const resetBoard: (count: number) => number[]  = (count) => {
  const numbers = Array.from({length: count}, (_, i) => i)
  return numbers.concat(numbers).sort(() => Math.random() - 0.5)
}
