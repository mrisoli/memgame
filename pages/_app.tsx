import App from 'next/app'
import { BoardProvider } from '../lib/context'

class MyApp extends App {
  render() {
    const { Component, pageProps } = this.props
    return (
      <BoardProvider>
        <Component {...pageProps} />
      </BoardProvider>
    )
  }
}

export default MyApp
