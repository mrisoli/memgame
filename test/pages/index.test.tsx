import React from 'react'
import { screen, RenderOptions, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { render } from '../testUtils'
import { Home } from '../../pages/index'
import { BoardProvider } from '../../lib/context'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

const renderWithContext = (options: RenderOptions) =>
  render(
    <BoardProvider>
      <Home />
    </BoardProvider>,
    options
  )

describe('Home page', () => {
  const randomBackup = global.Math.random
  beforeAll(() => {
    global.Math.random = () => 1
  })

  afterAll(() => {
    global.Math.random = randomBackup
  })

  describe('with default values', () => {
    beforeEach(() => {
      useRouter.mockReturnValue({ query: {} })
    })

    afterEach(() => {
      useRouter.mockClear()
    })

    it('matches snapshot', () => {
      const { asFragment } = renderWithContext({})
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders a number of cards equal to count', () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      expect(cards).toHaveLength(12)
    })

    it('renders a game in progress', () => {
      const { getByText } = renderWithContext({})
      expect(getByText('IN PROGRESS')).toBeTruthy()
    })
  })

  describe('with query parameters', () => {
    beforeEach(() => {
      useRouter.mockReturnValue({ query: { count: '2' } })
    })

    afterEach(() => {
      useRouter.mockClear()
    })

    it('renders a game in progress', () => {
      const { getByText } = renderWithContext({})
      expect(getByText('IN PROGRESS')).toBeTruthy()
    })

    it('renders a number of cards equal to count', () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      expect(cards).toHaveLength(4)
    })

    it('shows card after click', () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      userEvent.click(cards[0])
      expect(screen.getByText('0')).toBeTruthy()
    })

    it('wins game after all cards found', () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      userEvent.click(cards[0])
      userEvent.click(cards[2])
      userEvent.click(cards[1])
      userEvent.click(cards[3])
      expect(screen.getByText('YOU WIN!')).toBeTruthy()
    })

    it('closes cards after non matching pair', async () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      userEvent.click(cards[0])
      userEvent.click(cards[1])
      await waitFor(() => expect(screen.queryByText('0')).toBeFalsy())
    })

    it('resets board on reset click', () => {
      renderWithContext({})
      const cards = screen.getAllByTestId('card')
      const resetButton = screen.getByText('Reset')
      userEvent.click(cards[0])
      userEvent.click(resetButton)
      expect(screen.queryByText('0')).toBeFalsy()
    })
  })
})
