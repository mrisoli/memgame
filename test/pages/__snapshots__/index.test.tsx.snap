// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-247033640 container"
  >
    <main
      class="jsx-247033640"
    >
      <div
        class="jsx-247033640"
      >
        Board
      </div>
      <div
        class="jsx-247033640"
      >
        Status
      </div>
      <div
        class="jsx-247033640"
      >
        Reset
      </div>
    </main>
  </div>
</DocumentFragment>
`;
